csrc/multidevice/utils.cpp

```diff
@@ -13,22 +13,93 @@
 #include <multidevice/lower_communication.h>
 #include <multidevice/utils.h>
 #include <ops/all_ops.h>
+#include <root_domain_map.h>
 #include <scheduler/utils.h>
 
 #include <c10/util/irange.h>
 
 namespace nvfuser {
 namespace {
 
-std::vector<IterDomain*> getShardedIterDomains(TensorView* tv) {
-  std::vector<IterDomain*> sharded_ids;
+const std::unordered_set<IterDomain*> getShardedIterDomains(TensorView* tv) {
+  std::unordered_set<IterDomain*> sharded_ids;
   std::copy_if(
       tv->getLeafDomain().begin(),
       tv->getLeafDomain().end(),
-      std::back_inserter(sharded_ids),
+      std::inserter(sharded_ids, sharded_ids.begin()),
       [](auto id) { return id->isDeviceDim(); });
   return sharded_ids;
 }
+
+// Returns whether a IterDomain in a TensorView can be read/written
+// contiguously. when a IterDomain's sharding is flipped. This occurs when the
+// IterDomain being unsharded/sharded is the outermost in the allocation domain.
+// e.g. [DIDx(i0), r1, i2, i3]
+// i0 => true, outermost axis
+// i1 => true, reduction axis are not allocated
+// i2 => true, reduction and device axis are not allocated
+// i3 => false, i2 is allocated.
+bool isContiguousShard(TensorView* tv, IterDomain* changed_id) {
+  bool not_allocated = true;
+  for (auto id : tv->getLeafDomain()) {
```

Comment:
I assume this is another place that we'll have to fix later when allocation domain supports multidevice better? I'm curious what supports are missing. Do you have a bug about it? I'm sure @jjsjann123 is all ears for this.

Reply:
Yes, this helper function will need to be updated. imo the allocation domain needs to support multidevice better when we allow sharding on merged/split axis. 

Some background: Currently, sharding is only the RFactor domain. We have the constraint that `extent of a sharded axis == number of devices`, which guarantees the `sharded size of the axis = 1` or the actual allocation size of that axis = 1.

Eventually, we want to shard a split axis:
```
D = # of devices contained in DIDx
tv = makeContigTensor(2); [i0, i1]
tv->split(0, D, outer=true); //[D, i0//D, i1]
tv->axis(0)->parallelize(DIDx);
```
The allocation domain should be [i0//D, i1]
I might be wrong, but since we don't explicitly set the allocation domain for a tv with device parallelization (maybe we should?), the allocation domain is inferred from the rfactor domain?

As an aside sharding a split axis makes a lot of sense to me...but I don't know whether we should allow sharding a merged axis...

Reply:
As part of indexing refactoring, I'll probably explicitly set the allocation domain for all tensors. 

> As an aside sharding a split axis makes a lot of sense to me...but I don't know whether we should allow sharding a merged axis...

Maybe not common, but we may do that with `BIDx` as long as the extent is smaller than the maximum `gridDim.x`. Probably unlikely for `DIDx`.